// angular-instantsearch/esm2015/highlight/highlight.js
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { highlight } from 'instantsearch.js/es/helpers';
import { getPropertyByPath } from 'instantsearch.js/es/lib/utils';
export class NgAisHighlight {
    constructor() {
        this.tagName = 'mark';
    }
    get content() {
        const highlightAttributeResult = getPropertyByPath(this.hit._highlightResult, this.attribute);
        const fallback = getPropertyByPath(this.hit, this.attribute);
        // @MAJOR drop this custom fallback once it is implemented directly in instantsearch.js v5
        if (!highlightAttributeResult && fallback) {
            return fallback;
        }
        return highlight({
            attribute: this.attribute,
            highlightedTagName: this.tagName,
            hit: this.hit,
        });
    }
}
NgAisHighlight.decorators = [
    { type: Component, args: [{
                selector: 'ais-highlight',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: `<span class="ais-Highlight" [innerHtml]="content"></span>`
            },] }
];
NgAisHighlight.propDecorators = {
    attribute: [{ type: Input }],
    hit: [{ type: Input }],
    tagName: [{ type: Input }]
};
