// angular-instantsearch/esm2015/query-rule-custom-data/query-rule-custom-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, TemplateRef, Inject, forwardRef, Input, } from '@angular/core';
import { connectQueryRules } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
export class NgAisQueryRuleCustomData extends BaseWidget {
    /**
     * @param {?} instantSearchParent
     */
    constructor(instantSearchParent) {
        super('QueryRuleCustomData');
        this.instantSearchParent = instantSearchParent;
        this.state = {
            items: [],
        };
    }
    /**
     * @return {?}
     */
    get templateContext() {
        return {
            items: this.state.items,
        };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.createWidget(connectQueryRules, {
            transformItems: this.transformItems,
        });
        super.ngOnInit();
    }
}
NgAisQueryRuleCustomData.decorators = [
    { type: Component, args: [{
                selector: 'ais-query-rule-custom-data',
                template: `
    <div [class]="cx()">
      <ng-container *ngTemplateOutlet="template; context: templateContext">
      </ng-container>

      <div *ngIf="!template">
        <div *ngFor="let item of state.items">
          <pre>{{ item | json }}</pre>
        </div>
      </div>
    </div>
  `
            }] }
];
/** @nocollapse */
NgAisQueryRuleCustomData.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => NgAisInstantSearch)),] }] }
];
NgAisQueryRuleCustomData.propDecorators = {
    template: [{ type: ContentChild, args: [TemplateRef,] }],
    transformItems: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.template;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.transformItems;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.state;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.instantSearchParent;
}
