// angular-instantsearch/esm2015/query-rule-context/query-rule-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Inject, forwardRef } from '@angular/core';
import { connectQueryRules } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
export class NgAisQueryRuleContext extends BaseWidget {
    /**
     * @param {?} instantSearchParent
     */
    constructor(instantSearchParent) {
        super('QueryRuleContext');
        this.instantSearchParent = instantSearchParent;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.createWidget(connectQueryRules, {
            trackedFilters: this.trackedFilters,
            transformRuleContexts: this.transformRuleContexts,
        });
        super.ngOnInit();
    }
}
NgAisQueryRuleContext.decorators = [
    { type: Component, args: [{
                selector: 'ais-query-rule-context',
                template: ''
            }] }
];
/** @nocollapse */
NgAisQueryRuleContext.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => NgAisInstantSearch)),] }] }
];
NgAisQueryRuleContext.propDecorators = {
    trackedFilters: [{ type: Input }],
    transformRuleContexts: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgAisQueryRuleContext.prototype.trackedFilters;
    /** @type {?} */
    NgAisQueryRuleContext.prototype.transformRuleContexts;
    /** @type {?} */
    NgAisQueryRuleContext.prototype.instantSearchParent;
}
