// angular-instantsearch/esm5/query-rule-context/query-rule-context.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Inject, forwardRef } from '@angular/core';
import { connectQueryRules } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
var NgAisQueryRuleContext = /** @class */ (function (_super) {
    tslib_1.__extends(NgAisQueryRuleContext, _super);
    function NgAisQueryRuleContext(instantSearchParent) {
        var _this = _super.call(this, 'QueryRuleContext') || this;
        _this.instantSearchParent = instantSearchParent;
        return _this;
    }
    /**
     * @return {?}
     */
    NgAisQueryRuleContext.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createWidget(connectQueryRules, {
            trackedFilters: this.trackedFilters,
            transformRuleContexts: this.transformRuleContexts,
        });
        _super.prototype.ngOnInit.call(this);
    };
    NgAisQueryRuleContext.decorators = [
        { type: Component, args: [{
                    selector: 'ais-query-rule-context',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    NgAisQueryRuleContext.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return NgAisInstantSearch; })),] }] }
    ]; };
    NgAisQueryRuleContext.propDecorators = {
        trackedFilters: [{ type: Input }],
        transformRuleContexts: [{ type: Input }]
    };
    return NgAisQueryRuleContext;
}(BaseWidget));
export { NgAisQueryRuleContext };
if (false) {
    /** @type {?} */
    NgAisQueryRuleContext.prototype.trackedFilters;
    /** @type {?} */
    NgAisQueryRuleContext.prototype.transformRuleContexts;
    /** @type {?} */
    NgAisQueryRuleContext.prototype.instantSearchParent;
}
