// angular-instantsearch/esm5/query-rule-custom-data/query-rule-custom-data.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ContentChild, TemplateRef, Inject, forwardRef, Input, } from '@angular/core';
import { connectQueryRules } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
var NgAisQueryRuleCustomData = /** @class */ (function (_super) {
    tslib_1.__extends(NgAisQueryRuleCustomData, _super);
    function NgAisQueryRuleCustomData(instantSearchParent) {
        var _this = _super.call(this, 'QueryRuleCustomData') || this;
        _this.instantSearchParent = instantSearchParent;
        _this.state = {
            items: [],
        };
        return _this;
    }
    Object.defineProperty(NgAisQueryRuleCustomData.prototype, "templateContext", {
        get: /**
         * @return {?}
         */
        function () {
            return {
                items: this.state.items,
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgAisQueryRuleCustomData.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createWidget(connectQueryRules, {
            transformItems: this.transformItems,
        });
        _super.prototype.ngOnInit.call(this);
    };
    NgAisQueryRuleCustomData.decorators = [
        { type: Component, args: [{
                    selector: 'ais-query-rule-custom-data',
                    template: "\n    <div [class]=\"cx()\">\n      <ng-container *ngTemplateOutlet=\"template; context: templateContext\">\n      </ng-container>\n\n      <div *ngIf=\"!template\">\n        <div *ngFor=\"let item of state.items\">\n          <pre>{{ item | json }}</pre>\n        </div>\n      </div>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    NgAisQueryRuleCustomData.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return NgAisInstantSearch; })),] }] }
    ]; };
    NgAisQueryRuleCustomData.propDecorators = {
        template: [{ type: ContentChild, args: [TemplateRef,] }],
        transformItems: [{ type: Input }]
    };
    return NgAisQueryRuleCustomData;
}(BaseWidget));
export { NgAisQueryRuleCustomData };
if (false) {
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.template;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.transformItems;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.state;
    /** @type {?} */
    NgAisQueryRuleCustomData.prototype.instantSearchParent;
}
